rust/lance-encoding/src/encodings/physical/bitpack.rs

```diff
@@ -0,0 +1,531 @@
+// bitpacking encoding. Returns `None` if the type or array data is not supported.
+pub fn num_compressed_bits(arr: ArrayRef) -> Option<u64> {
+    match arr.data_type() {
+        DataType::UInt8 => num_bits_for_type::<UInt8Type>(arr.as_primitive()),
+        DataType::UInt16 => num_bits_for_type::<UInt16Type>(arr.as_primitive()),
+        DataType::UInt32 => num_bits_for_type::<UInt32Type>(arr.as_primitive()),
+        DataType::UInt64 => num_bits_for_type::<UInt64Type>(arr.as_primitive()),
+        _ => None,
+    }
+}
+
+// Compute the number of bits to to use for bitpacking generically.
+// Returns `None` if the array is empty or only contains null values.
+fn num_bits_for_type<T>(arr: &PrimitiveArray<T>) -> Option<u64>
+where
+    T: ArrowPrimitiveType,
+    T::Native: PrimInt + AsPrimitive<u64>,
+{
+    let max = arrow::compute::bit_or(arr);
+    max.map(|x| arr.data_type().byte_width() as u64 * 8 - x.leading_zeros() as u64)
+}
+
+#[derive(Debug, Default)]
+pub struct BitpackingBufferEncoder {}
+
+impl BufferEncoder for BitpackingBufferEncoder {
+    fn encode(&self, arrays: &[ArrayRef]) -> Result<EncodedBuffer> {
+        // TODO need to pass the num bits in 
+        let mut num_bits = 0;
+        for arr in arrays {
+            let arr_max = num_compressed_bits(arr.clone()).unwrap(); // TODO nounwrap
+            num_bits = num_bits.max(arr_max);
+        }
+
+        // TODO handle case where there are no arrays or all the arrays are full of zeros
+        if num_bits == 0 {
+            panic!("TODO handle zero length bitpacking")
+        }
+
+        let mut packed_arrays = vec![];
+        for arr in arrays {
+            let packed = pack_array(arr.clone(), num_bits)?;
+            packed_arrays.push(packed.into());
+        }
+
+        let data_type = arrays[0].data_type();
+        Ok(EncodedBuffer {
+            bits_per_value: num_bits,
+            parts: packed_arrays,
+        })
+    }
+}
+
+fn pack_buffers(data: ArrayData, num_bits: u64, byte_len: usize) -> Vec<u8> {
+    let buffers = data.buffers();
+    let mut packed_buffers = vec![];
+    for buffer in buffers {
+        for i in 0..800 {
+            if i % 4 == 0 {
+                println!("{} = {}", i / 4, &buffer[i]);
+            }
+        }
+        let packed_buffer = pack_bits(&buffer, num_bits, byte_len);
+        packed_buffers.push(packed_buffer);
+    }
+    packed_buffers.concat()
+}
+
+fn pack_bits(src: &[u8], num_bits: u64, byte_len: usize) -> Vec<u8> {
+    // calculate the total number of bytes we need to allocate for the destination.
+    // this will be the number of items in the source array times the number of bits.
+    let src_items = src.len() / byte_len as usize;
+    let mut dst_bytes_total = src_items * num_bits as usize / 8;
+
+    // if if there's a partial byte at the end, we need to allocate one more byte
+    if (src_items * num_bits as usize) % 8 != 0 {
+        dst_bytes_total += 1;
+    }
```

Comment:
can we do a divide_round_up here? like `(src_items * num_bits as usize + 7) / 8;`

Reply:
good call. I ended up using https://docs.rs/arrow/latest/arrow/util/bit_util/fn.ceil.html